File: source/_codes/java_review/HeapSort.java
package vell.bibi.sort_algorithms;

import vell.bibi.sort_algorithms.util.vell001;

public class HeapSort {

	public static void heapAdjust(int[] a, int father, int length){
		int child, cup;
		for(child=father*2+1, cup=a[father]; child<length; father=child, child=father*2+1){
			if(child+1 < length && a[child+1] > a[child])
				child ++;
			if(a[child] > cup){
				a[father] = a[child];
				a[child] = cup;
			} else
				break;
		}
	}
	
	public static void sort(int[] a){
		int cup;
		for(int i=a.length/2; i>=0; i--){
			heapAdjust(a, i, a.length);
		}
		
		for(int i=a.length-1; i>0; i--){
			cup = a[0];
			a[0] = a[i];
			a[i] = cup;
			heapAdjust(a, 0, i);
		}
	}
	
	public static void main(String[] args) {
		int[] a = vell001.getRandomList(10, 100);
		vell001.printList(a);
		sort(a);
		vell001.printList(a);
	}

}

File: source/_codes/c_learning/heap_sort.c
#include "vell001.h"

#define NUM 15

// 保证i位置为堆中最大的节点
// a:待调整堆数组，n:数组长度, i:待调整元素位置
int* HeapAdjust(int* a, int n, int i){
	int child, cup;
	// i: 父节点位置，child: 子节点位置（左:2*i+1,右:2*i+2）
	for(cup = a[i], child=2*i+1; child<n; i=child, child=2*i+1){
		// 如果右节点大于左节点child指向右节点
		if(child < n-1 && a[child+1] > a[child]) child++;
		// 如果子节点大于父节点则交换，并继续查找下一个节点（把当前的子节点当成新的父节点）
		if(a[child] > cup){
			a[i] = a[child];
			a[child] = cup;
		} else
			break; //如果不大于，则退出
	}
	return a;
}

// a:待调整堆数组，n:数组长度
int* HeapSort(int* a, int n){
	int cup, i;
	// 初始化堆，即对所有的非叶子节点进行HeapAdjust
	for(i = n/2; i>=0; i--){
		HeapAdjust(a, n, i);
	}
	// 把堆顶节点（最大节点）和堆中最后一个节点交换，然后对堆中除最后一个节点外的其他节点进行HeapAdjust
	for(i = n-1; i>0; i--){
		cup = a[i];
		a[i] = a[0];
		a[0] = cup;
		HeapAdjust(a, i, 0);
	}
	return a;
}

void main(){
	int* a = GetRandomNum(NUM);
	PrintList(a, NUM);
	HeapSort(a, NUM);
	PrintList(a, NUM);
}
